engine/src/dskmac.cpp

```diff
@@ -5540,9 +5540,9 @@ static bool startprocess_write_uint32_to_fd(int fd, uint32_t value)
 
 static bool startprocess_write_cstring_to_fd(int fd, char *string)
 {
-	if (!startprocess_write_uint32_to_fd(fd, strlen(string) + 1))
+	if (!startprocess_write_uint32_to_fd(fd, uint32_t(strlen(string) + 1)))
 		return false;
-	if (write(fd, string, strlen(string) + 1) != strlen(string) + 1)
+    if (write(fd, string, strlen(string) + 1) != (ssize_t)strlen(string) + 1)
```

Comment:
There are two problems here that adding these type casts have exposed

- What happens when trying to write a Very Big String on a 64-bit system? `strlen(string) + 1` could overflow a `uint32_t`, in which case I would expect the function to return `false`
- Computing `strlen(string)` three times :sob:
- The only negative value that `write()` should ever return is `-1`.  I'd usually recommend checking the failure state (negative return value) and checking the length written as two separate expressions -- and casting the returned length written to `size_t` rather than the other way round.

Here's a vague attempt at sanification of this:

```c++
static bool startprocess_write_cstring_to_fd(int fd, char *string)
{
    size_t t_len = strlen(string) + 1;
    if (len > UINT32_MAX)
        return false;

    if (!startprocess_write_uint32_t_fd(fd, uint32_t(t_len)))
        return false;

    ssize_t t_status = write(fd, string, t_len);

    /* This verifies that t_status is positive, at which point you _know_
     * that it will fit into a size_t losslessly */
    return (t_status >= 0 && size_t(t_status) == t_len);
}
```